compiler/src/emit/native.rs

```diff
@@ -0,0 +1,50 @@
+use crate::bytecode::{Bytecode, Opcode};
+use crate::constant_pool::ConstantPool;
+use crate::emit::{emit, EmissionState};
+use analyzer::relations::NativeId;
+use analyzer::types::hir::TypedExpr;
+
+/// Emits a native sequence of instructions.
+pub(crate) fn emit_native(
```

Comment:
```suggestion
pub(crate) fn emit_intrinsics(
```

Reply:
for me natives are more VM's implementations, and intrinsics are compiler-defined functions

Reply:
An intrinsic function can be implemented directly by the compiler if possible, instead of calling an actual function in a library. An intrinsic function could have two implementations, one that is more performant but enabled only if optimizations are required. See this [definition](https://en.wikipedia.org/wiki/Intrinsic_function):
> Compilers that implement intrinsic functions generally enable them only when a program requests [optimization](https://en.wikipedia.org/wiki/Optimizing_compiler), otherwise falling back to a default implementation provided by the language [runtime system](https://en.wikipedia.org/wiki/Runtime_system) (environment).

Here are emitted native operations, that can only exist in the host VM. They expose C functions that could not be reached if they didn't exist. Adding two ints cannot be done without reaching the host language, here's why this function should be called *native*.

Reply:
When emitting for example an addition between two ints, The compiler will replace the call by compiler-defined code, which is simply one opcode, the compiler is only bound to its instruction set, and the fact that the vm _interprets_ IADD opcode does not makes `IADD` a kind of native function, moreover the compiler will not fallback to functions implemented in the _native_ language for those binary operators because there are instructions for them which makes the 'fallback' case unreachable.
binary operators are bytecode-level instructions, not native code, and the compiler will detect those methods _intrinsics_ methods to inline them with their bytecode equivalent

Reply:
The compiler cannot replace the call, since it cannot be user-defined. Intrinsic methods are methods with fallbacks, that the compiler could treat normally if it wanted to skip them. Here, because of the high level representation, the analyzer technically forgot that is a binary expression that cannot be defined as a function. The trick is to mark those functions as native to signal that the compiler must handle them and cannot emit a regular code. In fact, the whole pipeline is named as *natives* and not *intrinsics*.
Those are terms for the compiler, it doesn't matter if at the end the compiler have a dedicated opcode or just call a named regular function that the VM would need to intercept.

Reply:
Renamed to *primitives operations*.